File: src/applications/likwid-topology.c
/*
 * =======================================================================================
 *
 *      Filename:  likwid-topology.c
 *
 *      Description:  A application to determine the thread and cache topology
 *                    on x86 processors.
 *
 *      Version:   <VERSION>
 *      Released:  <DATE>
 *
 *      Author:  Jan Treibig (jt), jan.treibig@gmail.com
 *      Project:  likwid
 *
 *      Copyright (C) 2012 Jan Treibig 
 *
 *      This program is free software: you can redistribute it and/or modify it under
 *      the terms of the GNU General Public License as published by the Free Software
 *      Foundation, either version 3 of the License, or (at your option) any later
 *      version.
 *
 *      This program is distributed in the hope that it will be useful, but WITHOUT ANY
 *      WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR A
 *      PARTICULAR PURPOSE.  See the GNU General Public License for more details.
 *
 *      You should have received a copy of the GNU General Public License along with
 *      this program.  If not, see <http://www.gnu.org/licenses/>.
 *
 * =======================================================================================
 */
/* #####   HEADER FILE INCLUDES   ######################################### */
#include <stdlib.h>
#include <stdio.h>
#include <string.h>
#include <sched.h>
#include <sys/types.h>
#include <unistd.h>
#include <ctype.h>

#include <types.h>
#include <cpuid.h>
#include <timer.h>
#include <affinity.h>
#include <numa.h>
#include <cpuFeatures.h>
#include <tree.h>
#include <asciiBoxes.h>
#include <strUtil.h>

/* #####   MACROS  -  LOCAL TO THIS SOURCE FILE   ######################### */

#define HELP_MSG \
printf("\nlikwid-topology --  Version %d.%d \n\n",VERSION,RELEASE); \
printf("A tool to print the thread and cache topology on x86 CPUs.\n"); \
printf("Options:\n"); \
printf("-h\t Help message\n"); \
printf("-v\t Version information\n"); \
printf("-c\t list cache information\n"); \
printf("-C\t measure processor clock\n"); \
printf("-o\t Store output to file. (Optional: Apply text filter)\n"); \
printf("-g\t graphical output\n\n")

#define VERSION_MSG \
printf("likwid-topology  %d.%d \n\n",VERSION,RELEASE)

/* #####   FUNCTION DEFINITIONS  -  EXPORTED FUNCTIONS   ################## */

int main (int argc, char** argv)
{ 
    int optGraphical = 0;
    int optCaches = 0;
    int optClock = 0;
    int c;
    int tmp;
    TreeNode* socketNode;
    TreeNode* coreNode;
    TreeNode* threadNode;
    BoxContainer* container;
    bstring  argString;
    bstring  filterScript = bfromcstr("NO");
    FILE* OUTSTREAM = stdout;

    while ((c = getopt (argc, argv, "hvcCgo:")) != -1)
    {
        switch (c)
        {
            case 'h':
                HELP_MSG;
                exit (EXIT_SUCCESS);    
            case 'v':
                VERSION_MSG;
                exit (EXIT_SUCCESS);    
            case 'g':
                optGraphical = 1;
                break;
            case 'c':
                optCaches = 1;
                break;
            case 'C':
                optClock = 1;
                break;
            case 'o':
                if (! (argString = bSecureInput(200,optarg)))
                {
                    fprintf(stderr, "Failed to read argument string!\n");
                }

                OUTSTREAM = bstr_to_outstream(argString, filterScript);

                if(!OUTSTREAM)
                {
                    fprintf(stderr, "Failed to parse out file pattern.\n");
                }
                break;
            case '?':
                if (isprint (optopt))
                {
                    fprintf (stderr, "Unknown option `-%c'.\n", optopt);
                }
                else
                {
                    fprintf (stderr,
                            "Unknown option character `\\x%x'.\n",
                            optopt);
                }
                return EXIT_FAILURE;
            default:
                HELP_MSG;
                exit (EXIT_SUCCESS);    
        }
    }

    cpuid_init();
    affinity_init();
    numa_init();

    fprintf(OUTSTREAM, HLINE);
    fprintf(OUTSTREAM, "CPU type:\t%s \n",cpuid_info.name);

    if (optClock)
    {
        timer_init();
        fprintf(OUTSTREAM, "CPU clock:\t%3.2f GHz \n",  (float) timer_getCpuClock() * 1.E-09);
    }

    /*----------------------------------------------------------------------
     *  Thread Topology
     *----------------------------------------------------------------------*/
    fprintf(OUTSTREAM, SLINE);
    fprintf(OUTSTREAM, "Hardware Thread Topology\n");
    fprintf(OUTSTREAM, SLINE);
    fprintf(OUTSTREAM, "Sockets:\t%u \n", cpuid_topology.numSockets);
    fprintf(OUTSTREAM, "Cores per socket:\t%u \n", cpuid_topology.numCoresPerSocket);
    fprintf(OUTSTREAM, "Threads per core:\t%u \n", cpuid_topology.numThreadsPerCore);
    fprintf(OUTSTREAM, HLINE);
    fprintf(OUTSTREAM, "HWThread\tThread\t\tCore\t\tSocket\n");

    for ( uint32_t i=0; i <  cpuid_topology.numHWThreads; i++)
    {
        fprintf(OUTSTREAM, "%d\t\t%u\t\t%u\t\t%u\n",i
                ,cpuid_topology.threadPool[i].threadId
                ,cpuid_topology.threadPool[i].coreId
                ,cpuid_topology.threadPool[i].packageId);
    }
    fprintf(OUTSTREAM, HLINE);

    socketNode = tree_getChildNode(cpuid_topology.topologyTree);
    while (socketNode != NULL)
    {
        fprintf(OUTSTREAM, "Socket %d: ( ",socketNode->id);
        coreNode = tree_getChildNode(socketNode);

        while (coreNode != NULL)
        {
            threadNode = tree_getChildNode(coreNode);

            while (threadNode != NULL)
            {
                fprintf(OUTSTREAM, "%d ",threadNode->id);
                threadNode = tree_getNextNode(threadNode);
            }
            coreNode = tree_getNextNode(coreNode);
        }
        socketNode = tree_getNextNode(socketNode);
        fprintf(OUTSTREAM, ")\n");
    }
    fprintf(OUTSTREAM, HLINE"\n");

    /*----------------------------------------------------------------------
     *  Cache Topology
     *----------------------------------------------------------------------*/
    fprintf(OUTSTREAM, SLINE);
    fprintf(OUTSTREAM, "Cache Topology\n");
    fprintf(OUTSTREAM, SLINE);

    for ( uint32_t i=0; i <  cpuid_topology.numCacheLevels; i++)
    {
        if (cpuid_topology.cacheLevels[i].type != INSTRUCTIONCACHE)
        {
            fprintf(OUTSTREAM, "Level:\t%d\n",cpuid_topology.cacheLevels[i].level);
            if (cpuid_topology.cacheLevels[i].size < 1048576)
            {
                fprintf(OUTSTREAM, "Size:\t%d kB\n",
                        cpuid_topology.cacheLevels[i].size/1024);
            }
            else 
            {
                fprintf(OUTSTREAM, "Size:\t%d MB\n",
                        cpuid_topology.cacheLevels[i].size/1048576);
            }

            if( optCaches)
            {
                switch (cpuid_topology.cacheLevels[i].type) {
                    case DATACACHE:
                        fprintf(OUTSTREAM, "Type:\tData cache\n");
                        break;

                    case INSTRUCTIONCACHE:
                        fprintf(OUTSTREAM, "Type:\tInstruction cache\n");
                        break;

                    case UNIFIEDCACHE:
                        fprintf(OUTSTREAM, "Type:\tUnified cache\n");
                        break;
                    default:
                        /* make the compiler happy */
                        break;
                }
                fprintf(OUTSTREAM, "Associativity:\t%d\n",
                        cpuid_topology.cacheLevels[i].associativity);
                fprintf(OUTSTREAM, "Number of sets:\t%d\n",
                        cpuid_topology.cacheLevels[i].sets);
                fprintf(OUTSTREAM, "Cache line size:%d\n",
                        cpuid_topology.cacheLevels[i].lineSize);
                if(cpuid_topology.cacheLevels[i].inclusive)
                {
                    fprintf(OUTSTREAM, "Non Inclusive cache\n");
                }
                else
                {
                    fprintf(OUTSTREAM, "Inclusive cache\n");
                }
                fprintf(OUTSTREAM, "Shared among %d threads\n",
                        cpuid_topology.cacheLevels[i].threads);
            }
            fprintf(OUTSTREAM, "Cache groups:\t");
            tmp = cpuid_topology.cacheLevels[i].threads;
            socketNode = tree_getChildNode(cpuid_topology.topologyTree);
            fprintf(OUTSTREAM, "( ");
            while (socketNode != NULL)
            {
                coreNode = tree_getChildNode(socketNode);

                while (coreNode != NULL)
                {
                    threadNode = tree_getChildNode(coreNode);

                    while (threadNode != NULL)
                    {

                        if (tmp)
                        {
                            fprintf(OUTSTREAM, "%d ",threadNode->id);
                            tmp--;
                        }
                        else
                        {
                            fprintf(OUTSTREAM, ") ( %d ",threadNode->id);
                            tmp = cpuid_topology.cacheLevels[i].threads;
                            tmp--;
                        }

                        threadNode = tree_getNextNode(threadNode);
                    }
                    coreNode = tree_getNextNode(coreNode);
                }
                socketNode = tree_getNextNode(socketNode);
            }
            fprintf(OUTSTREAM, ")\n");

            fprintf(OUTSTREAM, HLINE);
        }
    }

    fprintf(OUTSTREAM, "\n");

    /*----------------------------------------------------------------------
     *  NUMA Topology
     *----------------------------------------------------------------------*/
    fprintf(OUTSTREAM, SLINE);
    fprintf(OUTSTREAM, "NUMA Topology\n");
    fprintf(OUTSTREAM, SLINE);

    if (numa_init() < 0)
    {
        fprintf(OUTSTREAM, "NUMA is not supported on this node!\n");
    }
    else
    {
        fprintf(OUTSTREAM, "NUMA domains: %d \n",numa_info.numberOfNodes);
        fprintf(OUTSTREAM, HLINE);

        for ( uint32_t i = 0; i < numa_info.numberOfNodes; i++)
        {
            fprintf(OUTSTREAM, "Domain %d:\n", i);
            fprintf(OUTSTREAM, "Processors: ");

            for ( int j = 0; j < numa_info.nodes[i].numberOfProcessors; j++)
            {
                fprintf(OUTSTREAM, " %d",numa_info.nodes[i].processors[j]);
            }
            fprintf(OUTSTREAM, "\n");

            fprintf(OUTSTREAM, "Relative distance to nodes: ");

            for ( int j = 0; j < numa_info.nodes[i].numberOfDistances; j++)
            {
                fprintf(OUTSTREAM, " %d",numa_info.nodes[i].distances[j]);
            }
            fprintf(OUTSTREAM, "\n");

            fprintf(OUTSTREAM, "Memory: %g MB free of total %g MB\n",
                    numa_info.nodes[i].freeMemory/1024.0, numa_info.nodes[i].totalMemory/1024.0);
            fprintf(OUTSTREAM, HLINE);
        }
    }
    fprintf(OUTSTREAM, "\n");

    /*----------------------------------------------------------------------
     *  Graphical topology
     *----------------------------------------------------------------------*/
    if(optGraphical)
    {
        int j;
        bstring  boxLabel = bfromcstr("0");

        fprintf(OUTSTREAM, SLINE);
        fprintf(OUTSTREAM, "Graphical:\n");
        fprintf(OUTSTREAM, SLINE);

        /* Allocate without instruction cache */
        if ( cpuid_info.family == P6_FAMILY || cpuid_info.family == MIC_FAMILY ) 
        {
            container = asciiBoxes_allocateContainer(
                    cpuid_topology.numCacheLevels,
                    cpuid_topology.numCoresPerSocket);
        }
        else
        {
            container = asciiBoxes_allocateContainer(
                    cpuid_topology.numCacheLevels+1,
                    cpuid_topology.numCoresPerSocket);
        }

        socketNode = tree_getChildNode(cpuid_topology.topologyTree);
        while (socketNode != NULL)
        {
            fprintf(OUTSTREAM, "Socket %d:\n",socketNode->id);
            j=0;
            coreNode = tree_getChildNode(socketNode);

            /* add threads */
            while (coreNode != NULL)
            {
                threadNode = tree_getChildNode(coreNode);
                tmp =0;

                while (threadNode != NULL)
                {
                    if (tmp > 0)
                    {
                        bformata(boxLabel,"  %d", threadNode->id);
                    }
                    else
                    {
                        boxLabel = bformat("%d",threadNode->id);
                    }
                    tmp++;
                    threadNode = tree_getNextNode(threadNode);
                }
                asciiBoxes_addBox(container, 0, j, boxLabel); 
                j++;
                coreNode = tree_getNextNode(coreNode);
            }

            /* add caches */
            {
                int columnCursor=0;
                int lineCursor=1;
                uint32_t sharedCores;
                int numCachesPerLevel;
                int cacheWidth;

                for ( uint32_t i=0; i < cpuid_topology.numCacheLevels; i++ )
                {
                    sharedCores = cpuid_topology.cacheLevels[i].threads /
                        cpuid_topology.numThreadsPerCore;

                    if (cpuid_topology.cacheLevels[i].type != INSTRUCTIONCACHE)
                    {
                        if ( sharedCores > cpuid_topology.numCoresPerSocket )
                        {
                            numCachesPerLevel = 1;
                        }
                        else
                        {
                            numCachesPerLevel =
                                cpuid_topology.numCoresPerSocket/sharedCores;
                        }

                        columnCursor=0;
                        for ( j=0; j < numCachesPerLevel; j++ )
                        {
                            if (cpuid_topology.cacheLevels[i].size < 1048576)
                            {
                                boxLabel = bformat("%dkB",
                                        cpuid_topology.cacheLevels[i].size/1024);
                            }
                            else 
                            {
                                boxLabel = bformat("%dMB",
                                        cpuid_topology.cacheLevels[i].size/1048576);
                            }

                            if (sharedCores > 1)
                            {
                                if (sharedCores > cpuid_topology.numCoresPerSocket)
                                {
                                    cacheWidth = cpuid_topology.numCoresPerSocket-1;
                                }
                                else
                                {
                                    cacheWidth = sharedCores-1;
                                }
                                asciiBoxes_addJoinedBox(
                                        container,
                                        lineCursor,
                                        columnCursor,
                                        columnCursor+cacheWidth,
                                        boxLabel); 

                                columnCursor += sharedCores;
                            }
                            else 
                            {
                                asciiBoxes_addBox(
                                        container,
                                        lineCursor,
                                        columnCursor,
                                        boxLabel); 

                                columnCursor++;
                            }

                        }
                        lineCursor++;
                    }
                }
            }

            asciiBoxes_print(container);
            socketNode = tree_getNextNode(socketNode);
        }
        bdestroy(boxLabel);
    }

    fflush(OUTSTREAM);

    /* call filterscript if specified */
    if (!biseqcstr(filterScript,"NO"))
    {
        bcatcstr(filterScript, " topology");
        if (system(bdata(filterScript)) == EOF)
        {
            fprintf(stderr, "Failed to execute filter %s!\n", bdata(filterScript));
            exit(EXIT_FAILURE);
        }
    }

    return EXIT_SUCCESS;
}

File: trunk/src/applications/likwid-topology.lua
#!<PREFIX>/bin/likwid-lua

--[[
 * =======================================================================================
 *
 *      Filename:  likwid-topology.lua
 *
 *      Description:  A application to determine the thread and cache topology
 *                    on x86 processors.
 *
 *      Version:   <VERSION>
 *      Released:  <DATE>
 *
 *      Author:  Thomas Roehl (tr), thomas.roehl@gmail.com
 *      Project:  likwid
 *
 *      Copyright (C) 2014 Thomas Roehl
 *
 *      This program is free software: you can redistribute it and/or modify it under
 *      the terms of the GNU General Public License as published by the Free Software
 *      Foundation, either version 3 of the License, or (at your option) any later
 *      version.
 *
 *      This program is distributed in the hope that it will be useful, but WITHOUT ANY
 *      WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR A
 *      PARTICULAR PURPOSE.  See the GNU General Public License for more details.
 *
 *      You should have received a copy of the GNU General Public License along with
 *      this program.  If not, see <http://www.gnu.org/licenses/>.
 *
 * =======================================================================================]]
package.path = package.path .. ';<PREFIX>/share/lua/?.lua'

local likwid = require("likwid")
stdout_print = print

function version()
    print(string.format("likwid-topology --  Version %d.%d",likwid.version,likwid.release))
end

function usage()
    version()
    print("A tool to print the thread and cache topology on x86 CPUs.\n")
    print("Options:")
    print("-h, --help\t\t Help message")
    print("-v, --version\t\t Version information")
    print("-V, --verbose <level>\t Set verbosity")
    print("-c, --caches\t\t List cache information")
    print("-C, --clock\t\t Measure processor clock")
    print("-o, --output <file>\t Store output to file. (Optional: Apply text filter)")
    print("-g\t\t\t Graphical output")
end

print_caches = false
print_graphical = false
measure_clock = false
outfile = nil

for opt,arg in likwid.getopt(arg, {"h","v","c","C","g","o:","V:","help","version","verbose:","clock","caches","output:"}) do
    if (type(arg) == "string") then
        local s,e = arg:find("-");
        if s == 1 then
            print(string.format("Argmument %s to option -%s starts with invalid character -.", arg, opt))
            print("Did you forget an argument to an option?")
            os.exit(1)
        end
    end
    if opt == "h" or opt == "help" then
        usage()
        os.exit(0)
    elseif opt == "v" or opt == "version" then
        version()
        os.exit(0)
    elseif opt == "V" or opt == "verbose" then
        likwid.setVerbosity(tonumber(arg))
    elseif opt == "c" or opt == "caches" then
        print_caches = true
    elseif opt == "C" or opt == "clock" then
        measure_clock = true
    elseif opt == "g" then
        print_graphical = true
    elseif opt == "o" or opt == "output" then
        outfile = arg
        io.output(arg:gsub(string.match(arg, ".-[^\\/]-%.?([^%.\\/]*)$"),"tmp"))
        print = function(...) for k,v in pairs({...}) do io.write(v .. "\n") end end
    end
end

local config = likwid.getConfiguration()
local cpuinfo = likwid.getCpuInfo()
local cputopo = likwid.getCpuTopology()
local numainfo = likwid.getNumaInfo()
local affinity = likwid.getAffinityInfo()

print(likwid.hline)
print(string.format("CPU name:\t%s",cpuinfo["osname"]))
print(string.format("CPU type:\t%s",cpuinfo["name"]))
print(string.format("CPU stepping:\t%s",cpuinfo["stepping"]))
if (measure_clock) then
    if cpuinfo["clock"] == 0 then
        print(string.format("CPU clock:\t%3.2f GHz", likwid.getCpuClock() * 1.E-09))
    else
        print(string.format("CPU clock:\t%3.2f GHz", cpuinfo["clock"] * 1.E-09))
    end
end

print(likwid.sline)
print("Hardware Thread Topology")
print(likwid.sline)
print(string.format("Sockets:\t\t%u",cputopo["numSockets"]))
print(string.format("Cores per socket:\t%u",cputopo["numCoresPerSocket"]))
print(string.format("Threads per core:\t%u",cputopo["numThreadsPerCore"]))
print(likwid.hline)
print("HWThread\tThread\t\tCore\t\tSocket")

for cntr=0,cputopo["numHWThreads"]-1 do
    print(string.format("%d\t\t%u\t\t%u\t\t%u",cntr,
                        cputopo["threadPool"][cntr]["threadId"],
                        cputopo["threadPool"][cntr]["coreId"],
                        cputopo["threadPool"][cntr]["packageId"]))
end
print(likwid.hline)

for socket=0,cputopo["numSockets"]-1 do
    str = string.format("Socket %d: (",cputopo["topologyTree"][socket]["ID"])
    for core=0,cputopo["numCoresPerSocket"]-1 do
        for thread=0, cputopo["numThreadsPerCore"]-1 do
            str = str .. " " .. tostring(cputopo["topologyTree"][socket]["Childs"][core]["Childs"][thread])
        end
    end
    print(str .. " )")
end

print(likwid.hline .. "\n")

print(likwid.sline)
print("Cache Topology")
print(likwid.sline)

for level=1,cputopo["numCacheLevels"] do
    if (cputopo["cacheLevels"][level]["type"] ~= "INSTRUCTIONCACHE") then
        print(string.format("Level:\t%d",cputopo["cacheLevels"][level]["level"]))
        if (cputopo["cacheLevels"][level]["size"] < 1048576) then
            print(string.format("Size:\t%d kB",cputopo["cacheLevels"][level]["size"]/1024))
        else
            print(string.format("Size:\t%d MB",cputopo["cacheLevels"][level]["size"]/1048576))
        end
        
        if (print_caches) then
            if (cputopo["cacheLevels"][level]["type"] == "DATACACHE") then
                print("Type:\tData cache")
            elseif (cputopo["cacheLevels"][level]["type"] == "UNIFIEDCACHE") then
                print("Type:\tUnified cache")
            end

            print(string.format("Associativity:\t%d",cputopo["cacheLevels"][level]["associativity"]))
            print(string.format("Number of sets:\t%d",cputopo["cacheLevels"][level]["sets"]))
            print(string.format("Cache line size:%d",cputopo["cacheLevels"][level]["lineSize"]))
            
            if (cputopo["cacheLevels"][level]["inclusive"] > 0) then
                print("Non Inclusive cache")
            else
                print("Inclusive cache")
            end
            print(string.format("Shared among %d threads",cputopo["cacheLevels"][level]["threads"]))
        end
        local threads = cputopo["cacheLevels"][level]["threads"]
        str = "Cache groups:\t( "
        for socket=0,cputopo["numSockets"]-1 do
            for core=0,cputopo["numCoresPerSocket"]-1 do
                for cpu=0,cputopo["numThreadsPerCore"]-1 do
                    if (threads ~= 0) then
                        str = str .. cputopo["topologyTree"][socket]["Childs"][core]["Childs"][cpu] .. " "
                        threads = threads - 1
                    else
                        str = str .. string.format(") ( %d ",cputopo["topologyTree"][socket]["Childs"][core]["Childs"][cpu])
                        threads = cputopo["cacheLevels"][level]["threads"]
                        threads = threads - 1
                    end
                end
            end
        end
        print(str .. ")")
        print(likwid.hline)
    end
end
print("\n")

print(likwid.sline)
print("NUMA Topology")
print(likwid.sline)

if (numainfo["numberOfNodes"] == 0) then
    print("NUMA is not supported on this node!")
else
    print(string.format("NUMA domains: %d",numainfo["numberOfNodes"]))
    print(likwid.hline)
    -- -2 because numberOfNodes is seen as one entry
    for node=1,numainfo["numberOfNodes"] do
        print(string.format("Domain %d:",numainfo["nodes"][node]["id"]))
        str = "Processors: "
        for cpu=1,numainfo["nodes"][node]["numberOfProcessors"] do
            str = str .. " " .. numainfo["nodes"][node]["processors"][cpu]
        end
        print(str)
        str = "Relative distance to nodes: "
        for cpu=1,numainfo["nodes"][node]["numberOfDistances"] do
            str = str .. " " .. numainfo["nodes"][node]["distances"][cpu][cpu-1]
        end
        print(str)
        print(string.format("Memory: %g MB free of total %g MB", 
                                tonumber(numainfo["nodes"][node]["freeMemory"]/1024.0),
                                tonumber(numainfo["nodes"][node]["totalMemory"]/1024.0)))
        print(likwid.hline)
    end
end

if (print_graphical) then
    print("\n")
    print(likwid.sline)
    print("Graphical Topology")
    print(likwid.sline)
    for socket=0,cputopo["numSockets"]-1 do
        print(string.format("Socket %d:",cputopo["topologyTree"][socket]["ID"]))
        container = {}
        for core=0,cputopo["numCoresPerSocket"]-1 do
            local tmpString = ""
            for thread=0,cputopo["numThreadsPerCore"]-1 do
                if thread == 0 then
                    tmpString = tmpString .. tostring(cputopo["topologyTree"][socket]["Childs"][core]["Childs"][thread])
                else
                    tmpString = tmpString .. " " .. tostring(cputopo["topologyTree"][socket]["Childs"][core]["Childs"][thread]).. " "
                end
            end
            likwid.addSimpleAsciiBox(container, 1, core+1, tmpString)
        end
        
        local columnCursor = 1
        local lineCursor = 2
        for cache=1,cputopo["numCacheLevels"] do
            if cputopo["cacheLevels"][cache]["type"] ~= "INSTRUCTIONCACHE" then
                local cachesAtCurLevel = 0
                local sharedCores = cputopo["cacheLevels"][cache]["threads"]/cputopo["numThreadsPerCore"]
                if sharedCores >= cputopo["numCoresPerSocket"] then
                    cachesAtCurLevel = 1
                else
                    cachesAtCurLevel = cputopo["numCoresPerSocket"]/sharedCores
                end
                columnCursor = 1
                for cachesAtLevel=1,cachesAtCurLevel do
                    local tmpString = ""
                    local cacheWidth = 0
                    if cputopo["cacheLevels"][cache]["size"] < 1048576 then
                        tmpString = string.format("%dkB", cputopo["cacheLevels"][cache]["size"]/1024)
                    else
                        tmpString = string.format("%dMB", cputopo["cacheLevels"][cache]["size"]/1048576)
                    end
                    if sharedCores > 1 then
                        if sharedCores > cputopo["numCoresPerSocket"] then
                            cacheWidth = sharedCores
                        else
                            cacheWidth = sharedCores - 1
                        end
                        likwid.addJoinedAsciiBox(container, lineCursor, columnCursor,columnCursor + cacheWidth, tmpString)
                        columnCursor = columnCursor + cacheWidth
                    else
                        likwid.addSimpleAsciiBox(container, lineCursor, columnCursor, tmpString)
                        columnCursor = columnCursor + 1
                    end
                end
                lineCursor = lineCursor + 1
            end
        end
        likwid.printAsciiBox(container);
    end
end

if outfile then
    local suffix = string.match(outfile, ".-[^\\/]-%.?([^%.\\/]*)$")
    local command = "<PREFIX>/share/likwid/filter/" .. suffix 
    if suffix ~= "txt" then
        command = command .." ".. outfile:gsub("."..suffix,".tmp",1) .. " topology"
        local f = io.popen(command)
        local o = f:read("*a")
        if o:len() > 0 then
            print(string.format("Failed to executed filter script %s.",command))
        end
    else
        os.rename(outfile:gsub("."..suffix,".tmp",1), outfile)
    end
end

likwid.putAffinityInfo()
likwid.putNumaInfo()
likwid.putTopology()
likwid.putConfiguration()
